src/threaddeadlock.cpp

```diff
@@ -0,0 +1,626 @@
+// Copyright (c) 2019 Greg Griffith
+// Copyright (c) 2019 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "threaddeadlock.h"
+#include "util.h"
+
+#ifdef DEBUG_LOCKORDER // this ifdef covers the rest of the file
+
+// removes 1 lock for a critical section
+void _remove_lock_critical_exit(void *cs)
+{
+    if (!lockdata.available)
+    {
+        // lockdata was already deleted
+        return;
+    }
+    uint64_t tid = getTid();
+    auto it = lockdata.locksheldbythread.find(tid);
+    if (it == lockdata.locksheldbythread.end())
+    {
+        throw std::logic_error("unlocking non-existant lock");
+    }
+    if (it->second.back().first != cs)
+    {
+        LOGA("got %s but was not expecting it\n", it->second.back().second.ToString().c_str());
+        throw std::logic_error("unlock order inconsistant with lock order");
+    }
+    LockType type = it->second.back().second.GetLockType();
+    OwnershipType ownership = it->second.back().second.GetExclusive();
+    // assuming we unlock in the reverse order of locks, we can simply pop back
+    it->second.pop_back();
+    // if we have no more locks on this critical section...
+    if (type != LockType::SHARED_MUTEX)
+    {
+        for (auto entry : it->second)
+        {
+            if (entry.first == cs)
+            {
+                // we have another lock on this critical section
+                return;
+            }
+        }
+    }
+    // remove from the other maps
+    if (ownership == OwnershipType::EXCLUSIVE)
+    {
+        auto iter = lockdata.writelocksheld.find(cs);
+        if (iter != lockdata.writelocksheld.end())
+        {
+            if (iter->second.empty())
+                return;
+            if (iter->second.count(tid) != 0)
+            {
+                iter->second.erase(tid);
+            }
+        }
+    }
+    else // !isExclusive
+    {
+        auto iter = lockdata.readlocksheld.find(cs);
+        if (iter != lockdata.readlocksheld.end())
+        {
+            if (iter->second.empty())
+                return;
+            if (iter->second.count(tid) != 0)
+            {
+                iter->second.erase(tid);
+            }
+        }
+    }
+}
+
+void potential_deadlock_detected(LockStackEntry now, LockStack &deadlocks, std::set<uint64_t> &threads)
+{
+    LOGA("POTENTIAL DEADLOCK DETECTED\n");
```

Comment:
Its important to remember that a log is atomic, but multiple LOGA calls may be interspersed with other threads.  So To produce a single logged block, you'd construct the log inside a std::string and then issue a single call to LOGA.

That would be the perfect way to do it anyway.  IDK if we need perfect here because the system is going to abort due to a deadlock detection.  However, it may make sense to combine these 3 initial LOGA calls, or at least the last 2 so that "which has" (which provides  no context) doesn't end up in a line all by itself.